app/worker/worker.go

```diff
@@ -78,8 +86,16 @@ func InitAndRun(conf config.Reader, workflows Workflows, dataStore store.DataSto
 			break
 		}
 		switch msg := msg.(type) {
-		case *model.Job:
-			job := msg
+		case *natsio.Msg:
+			job := &model.Job{}
+			err := json.Unmarshal(msg.Data, job)
+			if err != nil {
+				l.Error(errors.Wrap(err, "failed to unmarshall message"))
+				if err := msg.Term(); err != nil {
+					l.Error(errors.Wrap(err, "failed to term the message"))
+				}
+				continue
+			}
```

Comment:
Term tells the server to not redeliver the message, we don't need to Ack the message in this case.
If a message is malformed and we cannot unmarshal it, we simply remove it from the queue.

Reply:
I meant ACK'ing the message after the message has been successfully unmarshalled. Now we don't ack the job untill it has been processed (after aquiring the semaphore) which increase the chance of the message being redelivered and processed multiple times.

Reply:
But that's the point of Ack: you want to do it once it's done; if the process gets terminated in the middle, the message will be re-delivered. Obviously we need to make sure we can complete the processing within the TTL, but our workflows are all very short.

Reply:
This will send InProgress messages if the job takes longer than 30 seconds to run:
https://github.com/mendersoftware/workflows/pull/173/commits/2032799c11bd0f34123fda347948361c691c4129